### ISSUE TITLE
ToC Generation

### ISSUE THREAD

**Issue by jdormit** [jdormit/livre#20]
From an email from @diethardsteiner:

> I had a look at how the ToC is generated (I must admit I am new to Electron
> and know only basics of HTML) and it seems like that for some reason the
> parent property value is never provided for the headings with the
> `getToc()` function issued in `index.html`? Also the `subitems` array seems
> to be constantly empty.
>
> I played around with a bare bones epub.js example and could extract this
> info. I was using the parent info to construct level info for the headings,
> so that later on some formatting can be applied. Simple example:
> 
> ```
> var tocBuilder = function(obj, trail = "root"){
> 
>   var currentNodeObj = {
>     trail: trail
>     , extractHeading: function(element){
>         var prnt = element.parent === undefined ? "" : ("/" +
> element.parent);
> 
>         var trail = this.trail + prnt;
>         var level = trail.split("/").length;
> 
>         console.log("Level: " + level + ", Trail: " + trail + ": " +
> element.label);
>         if(element.subitems.length > 0) tocBuilder(element.subitems, trail);
>       }
>   }
> 
>   obj.map(currentNodeObj.extractHeading, currentNodeObj)
> }
> 
> 
> tocBuilder(window.reader.book.toc)
> ```
> 
> I had hoped that getToc() returns also level info, which would allow a more
> elegant solution.
>

**Comment by jdormit** [comment]
Hi @diethardsteiner - were you able to see a ToC at all? I've seen inconsistent results - some eBooks have a ToC that `epub.js` can parse, and some seem not to, although it is entirely possible that I'm doing the parsing wrong. 

What do you mean by "the parent property value"?

**Comment by diethardsteiner** [comment]
@jdormit Yes, I can see a ToC. Ok, so now I opened the same book in Livre as in my simple epub.js/html example and I can see the `parent` element being populated (see attached screenshot), so yes, there seem to be inconsistencies with epubs. Hm, so I'll find another approach to determine the level.
![screenshot from 2017-05-05 18-05-57](https://cloud.githubusercontent.com/assets/1726057/25756248/f9904b70-31bd-11e7-813f-0cea4d328cdc.png)

**Comment by diethardsteiner** [comment]
@jdormit Ok, we can just use this instead:

```javascript
// since parent info is not available in all epubs, here an alt approach:
var tocBuilder = function(obj, level = 1){

  var currentNodeObj = {
    level: level
    , extractHeading: function(element){        
        console.log("Level: " + level + ", " + element.label);
        if(element.subitems.length > 0) tocBuilder(element.subitems, level + 1);
      }
  }

  obj.map(currentNodeObj.extractHeading, currentNodeObj)
} 

tocBuilder(window.reader.book.toc)
```

![screenshot from 2017-05-05 18-23-57](https://cloud.githubusercontent.com/assets/1726057/25756839/2aa539da-31c0-11e7-8da9-d084b27741fe.png)


This still only for testing. I'll try to add something similar to Livre over the next few days.

**Comment by diethardsteiner** [comment]
Ok, so it turns out Livre's ToC works fine with certain books (I just found one) ... so even indentation works. The issue seems to be that the other book I had a look at (where all levels of the TOC where displayed as the first level), for some reason, for all the elements in the ToC, no subitems were returned. This is also consistent with my simplified epub.js prototype. Then I also checked if iBooks can render the ToC of this books correctly, and indeed it does, so maybe there is a different structure in this epub file that epub.js doesn't pick up.

**Comment by diethardsteiner** [comment]
Maybe it is related to this [issue](https://github.com/futurepress/epub.js/issues/427).

**Comment by jdormit** [comment]
Just to clarify - on the first book you looked at, the table of contents was displayed, but all items were displayed as top-level items? Or did Livre display the "No table of contents available." message?

**Comment by diethardsteiner** [comment]
yes, the ToC was displayed for the first book, however, everything was displayed on the 1st level (and this seems to be caused by empty subitems).

**Comment by diethardsteiner** [comment]
I created [this issue](https://github.com/futurepress/epub.js/issues/615) - just as a reference.

**Comment by jdormit** [comment]
Perfect! I think we should wait and see if there is a response there. I just finished up my undergraduate degree, so I should have some free cycles pretty soon to give Livre some much needed love. Thank you for your interest, by the way!
